src/go/serviceconfig/service_config_fetcher.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceconfig
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/metadata"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+
+	confpb "google.golang.org/genproto/googleapis/api/serviceconfig"
+	smpb "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+)
+
+type ServiceConfigFetcher struct {
+	serviceName         string
+	checkRolloutsTicker *time.Ticker
+	client              http.Client
+	mf                  *metadata.MetadataFetcher
+	opts                options.ConfigGeneratorOptions
+
+	curServiceConfig    *confpb.Service
+	curRolloutId        string
+	curFetchConfigError error
+}
+
+func NewServiceConfigFetcher(mf *metadata.MetadataFetcher, opts options.ConfigGeneratorOptions, serviceName, configId string) (*ServiceConfigFetcher, error) {
+	caCert, err := ioutil.ReadFile(opts.RootCertsPath)
+	if err != nil {
+		return nil, err
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+	scf := &ServiceConfigFetcher{
+		client: http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs: caCertPool,
+				},
+			},
+			Timeout: opts.HttpRequestTimeout,
+		},
+		serviceName: serviceName,
+		mf:          mf,
+		opts:        opts,
+	}
+	_, _ = scf.fetchConfig(configId)
```

Comment:
Be careful about this, this makes it harder to unit test. Ideally, the constructor shouldn't call another function to make external requests like this. Instead, `configId` should be stored in the `struct` and the client should call `FetchConfig`.

Reply:
+1 
what is the purpose for this?